diff --git a/modules/components/org.wso2.intcloud.core/pom.xml b/modules/components/org.wso2.intcloud.core/pom.xml
index 53cf4b1..29f6873 100644
--- a/modules/components/org.wso2.intcloud.core/pom.xml
+++ b/modules/components/org.wso2.intcloud.core/pom.xml
@@ -74,6 +74,10 @@
             <groupId>org.wso2.appcloud</groupId>
             <artifactId>org.wso2.intcloud.services.tenant.tasks.stub</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.appcloud</groupId>
+            <artifactId>org.wso2.intcloud.services.tenant.ntasks.stub</artifactId>
+        </dependency>
     </dependencies>
 
     <build>
diff --git a/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/CarbonApplicationClient.java b/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/CarbonApplicationClient.java
index f6fb39e..55b391b 100644
--- a/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/CarbonApplicationClient.java
+++ b/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/CarbonApplicationClient.java
@@ -228,7 +228,7 @@ public class CarbonApplicationClient {
         String[] applicationList;
         try {
             applicationList = appAdminStub.listAllApplications();
-            log.info("Found applications " + applicationList);
+            log.info("Found applications " + Arrays.toString(applicationList));
         } catch (RemoteException | ApplicationAdminExceptionException e) {
             throw new IntCloudException("Error getting deployed application list from server", e);
         }
diff --git a/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ESBClient.java b/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ESBClient.java
index 10fc562..41c48ea 100644
--- a/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ESBClient.java
+++ b/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ESBClient.java
@@ -96,4 +96,8 @@ public class ESBClient {
             throws IntCloudException, TaskManagementException, XMLStreamException, IOException, InterruptedException {
         ScheduledTaskClient.getInstance().stopTask(tenantId, applicationName, taskConfiguration);
     }
+
+    public boolean getTaskStatus(int tenantId, String applicationName, String taskConfiguration) throws IntCloudException {
+        return ScheduledTaskClient.getInstance().getTaskStatus(tenantId, applicationName, taskConfiguration);
+    }
 }
diff --git a/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ScheduledTaskClient.java b/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ScheduledTaskClient.java
index c4f03b8..e1d7b7f 100644
--- a/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ScheduledTaskClient.java
+++ b/modules/components/org.wso2.intcloud.core/src/main/java/org/wso2/intcloud/core/client/ScheduledTaskClient.java
@@ -31,6 +31,8 @@ import org.wso2.carbon.task.stub.TaskAdminStub;
 import org.wso2.carbon.task.stub.TaskManagementException;
 import org.wso2.intcloud.common.IntCloudException;
 import org.wso2.intcloud.common.util.IntCloudUtil;
+import org.wso2.intcloud.services.tenant.ntasks.stub.types.TenantNTaskAdminServiceExceptionException;
+import org.wso2.intcloud.services.tenant.ntasks.stub.types.TenantNTaskAdminServiceStub;
 import org.wso2.intcloud.services.tenant.tasks.stub.types.TenantTaskAdminServiceStub;
 import org.wso2.intcloud.services.tenant.tasks.stub.types.TenantTaskAdminServiceTaskManagementExceptionException;
 
@@ -49,6 +51,8 @@ public class ScheduledTaskClient {
 
     TenantTaskAdminServiceStub tenantTaskAdminServiceStub = null;
 
+    TenantNTaskAdminServiceStub tenantNTaskAdminServiceStub = null;
+
     private ScheduledTaskClient() throws IntCloudException {
 
         try {
@@ -79,6 +83,22 @@ public class ScheduledTaskClient {
         ten_authenticator.setPreemptiveAuthentication(true);
         ten_client_options.setProperty(org.apache.axis2.transport.http.HTTPConstants.AUTHENTICATE, ten_authenticator);
         ten_client.setOptions(ten_client_options);
+
+        try {
+            tenantNTaskAdminServiceStub =
+                    new TenantNTaskAdminServiceStub(IntCloudUtil.getPropertyValue("TenantNTaskAdminService"));
+        } catch (AxisFault axisFault) {
+            throw new IntCloudException(axisFault.getMessage(), axisFault);
+        }
+        ServiceClient ten_ntask_client = tenantNTaskAdminServiceStub._getServiceClient();
+        Options ten_ntask_client_options = ten_ntask_client.getOptions();
+        HttpTransportProperties.Authenticator ten_ntask_authenticator = new HttpTransportProperties.Authenticator();
+        ten_ntask_authenticator.setUsername(IntCloudUtil.getPropertyValue("ESBServerUserName"));
+        ten_ntask_authenticator.setPassword(IntCloudUtil.getPropertyValue("ESBServerPassword"));
+        ten_ntask_authenticator.setPreemptiveAuthentication(true);
+        ten_ntask_client_options
+                .setProperty(org.apache.axis2.transport.http.HTTPConstants.AUTHENTICATE, ten_ntask_authenticator);
+        ten_ntask_client.setOptions(ten_ntask_client_options);
     }
 
     public static ScheduledTaskClient getInstance() throws IntCloudException {
@@ -98,8 +118,21 @@ public class ScheduledTaskClient {
         try {
             String taskConfiguration = getTaskConfiguration(applicationName, paramConfigurationJSON);
             taskOM = AXIOMUtil.stringToOM(taskConfiguration);
+
+            String taskName = taskOM.getAttributeValue(new QName("name"));
+            String taskGroup = taskOM.getAttributeValue(new QName("group"));
+
             log.info("Deploying task configuration : " + taskConfiguration);
+
+            boolean taskExists = tenantTaskAdminServiceStub.taskExists(tenantId, taskName, taskGroup);
+
+            if(taskExists) {
+                log.info("Deleting existing task " + taskName + " to redeploy");
+                tenantTaskAdminServiceStub.deleteTaskDescriptionInTenant(tenantId, taskName, taskGroup);
+            }
+
             tenantTaskAdminServiceStub.addTaskDescriptionInTenant(tenantId, taskConfiguration);
+
         } catch (XMLStreamException | RemoteException | TenantTaskAdminServiceTaskManagementExceptionException e) {
             throw new IntCloudException(e.getMessage(), e);
         }
@@ -256,4 +289,22 @@ public class ScheduledTaskClient {
 
         deleteTask(taskName, taskGroup);
     }
+
+    public boolean getTaskStatus(int tenantId, String applicationName, String taskConfiguration)
+            throws IntCloudException {
+        OMElement task = null;
+        try {
+            task = AXIOMUtil.stringToOM(taskConfiguration);
+        } catch (XMLStreamException e) {
+            throw new IntCloudException("Error getting status of task for " + applicationName, e);
+        }
+
+        String taskName = task.getAttributeValue(new QName("name"));
+
+        try {
+            return tenantNTaskAdminServiceStub.isESBTaskRunningInTenant(tenantId, taskName);
+        } catch (RemoteException | TenantNTaskAdminServiceExceptionException e) {
+            throw new IntCloudException("Error getting state of integration task for " + applicationName, e);
+        }
+    }
 }
\ No newline at end of file
diff --git a/modules/service-stubs/org.wso2.intcloud.services.tenant.tasks.stub/src/main/resources/TenantTaskAdminService.wsdl b/modules/service-stubs/org.wso2.intcloud.services.tenant.tasks.stub/src/main/resources/TenantTaskAdminService.wsdl
index 214baf9..3fbaea9 100644
--- a/modules/service-stubs/org.wso2.intcloud.services.tenant.tasks.stub/src/main/resources/TenantTaskAdminService.wsdl
+++ b/modules/service-stubs/org.wso2.intcloud.services.tenant.tasks.stub/src/main/resources/TenantTaskAdminService.wsdl
@@ -15,7 +15,7 @@
                </xs:sequence>
             </xs:complexType>
          </xs:element>
-         <xs:element name="deleteTaskDescriptionInTenant">
+         <xs:element name="taskExists">
             <xs:complexType>
                <xs:sequence>
                   <xs:element minOccurs="0" name="tenantId" type="xs:int"/>
@@ -24,7 +24,7 @@
                </xs:sequence>
             </xs:complexType>
          </xs:element>
-         <xs:element name="deleteTaskDescriptionInTenantResponse">
+         <xs:element name="taskExistsResponse">
             <xs:complexType>
                <xs:sequence>
                   <xs:element minOccurs="0" name="return" type="xs:boolean"/>
@@ -46,6 +46,22 @@
                </xs:sequence>
             </xs:complexType>
          </xs:element>
+         <xs:element name="deleteTaskDescriptionInTenant">
+            <xs:complexType>
+               <xs:sequence>
+                  <xs:element minOccurs="0" name="tenantId" type="xs:int"/>
+                  <xs:element minOccurs="0" name="name" nillable="true" type="xs:string"/>
+                  <xs:element minOccurs="0" name="group" nillable="true" type="xs:string"/>
+               </xs:sequence>
+            </xs:complexType>
+         </xs:element>
+         <xs:element name="deleteTaskDescriptionInTenantResponse">
+            <xs:complexType>
+               <xs:sequence>
+                  <xs:element minOccurs="0" name="return" type="xs:boolean"/>
+               </xs:sequence>
+            </xs:complexType>
+         </xs:element>
       </xs:schema>
    </wsdl:types>
    <wsdl:message name="deleteTaskDescriptionInTenantRequest">
@@ -57,6 +73,12 @@
    <wsdl:message name="TenantTaskAdminServiceTaskManagementException">
       <wsdl:part name="parameters" element="ns:TenantTaskAdminServiceTaskManagementException"/>
    </wsdl:message>
+   <wsdl:message name="taskExistsRequest">
+      <wsdl:part name="parameters" element="ns:taskExists"/>
+   </wsdl:message>
+   <wsdl:message name="taskExistsResponse">
+      <wsdl:part name="parameters" element="ns:taskExistsResponse"/>
+   </wsdl:message>
    <wsdl:message name="addTaskDescriptionInTenantRequest">
       <wsdl:part name="parameters" element="ns:addTaskDescriptionInTenant"/>
    </wsdl:message>
@@ -69,10 +91,14 @@
          <wsdl:output message="ns:deleteTaskDescriptionInTenantResponse" wsaw:Action="urn:deleteTaskDescriptionInTenantResponse"/>
          <wsdl:fault message="ns:TenantTaskAdminServiceTaskManagementException" name="TenantTaskAdminServiceTaskManagementException" wsaw:Action="urn:deleteTaskDescriptionInTenantTenantTaskAdminServiceTaskManagementException"/>
       </wsdl:operation>
+      <wsdl:operation name="taskExists">
+         <wsdl:input message="ns:taskExistsRequest" wsaw:Action="urn:taskExists"/>
+         <wsdl:output message="ns:taskExistsResponse" wsaw:Action="urn:taskExistsResponse"/>
+         <wsdl:fault message="ns:TenantTaskAdminServiceTaskManagementException" name="TenantTaskAdminServiceTaskManagementException" wsaw:Action="urn:taskExistsTenantTaskAdminServiceTaskManagementException"/>
+      </wsdl:operation>
       <wsdl:operation name="addTaskDescriptionInTenant">
          <wsdl:input message="ns:addTaskDescriptionInTenantRequest" wsaw:Action="urn:addTaskDescriptionInTenant"/>
          <wsdl:output message="ns:addTaskDescriptionInTenantResponse" wsaw:Action="urn:addTaskDescriptionInTenantResponse"/>
-         <wsdl:fault message="ns:TenantTaskAdminServiceTaskManagementException" name="TenantTaskAdminServiceTaskManagementException" wsaw:Action="urn:addTaskDescriptionInTenantTenantTaskAdminServiceTaskManagementException"/>
       </wsdl:operation>
    </wsdl:portType>
    <wsdl:binding name="TenantTaskAdminServiceSoap11Binding" type="ns:TenantTaskAdminServicePortType">
@@ -89,8 +115,8 @@
             <soap:fault use="literal" name="TenantTaskAdminServiceTaskManagementException"/>
          </wsdl:fault>
       </wsdl:operation>
-      <wsdl:operation name="addTaskDescriptionInTenant">
-         <soap:operation soapAction="urn:addTaskDescriptionInTenant" style="document"/>
+      <wsdl:operation name="taskExists">
+         <soap:operation soapAction="urn:taskExists" style="document"/>
          <wsdl:input>
             <soap:body use="literal"/>
          </wsdl:input>
@@ -101,6 +127,15 @@
             <soap:fault use="literal" name="TenantTaskAdminServiceTaskManagementException"/>
          </wsdl:fault>
       </wsdl:operation>
+      <wsdl:operation name="addTaskDescriptionInTenant">
+         <soap:operation soapAction="urn:addTaskDescriptionInTenant" style="document"/>
+         <wsdl:input>
+            <soap:body use="literal"/>
+         </wsdl:input>
+         <wsdl:output>
+            <soap:body use="literal"/>
+         </wsdl:output>
+      </wsdl:operation>
    </wsdl:binding>
    <wsdl:binding name="TenantTaskAdminServiceSoap12Binding" type="ns:TenantTaskAdminServicePortType">
       <soap12:binding transport="http://schemas.xmlsoap.org/soap/http" style="document"/>
@@ -116,8 +151,8 @@
             <soap12:fault use="literal" name="TenantTaskAdminServiceTaskManagementException"/>
          </wsdl:fault>
       </wsdl:operation>
-      <wsdl:operation name="addTaskDescriptionInTenant">
-         <soap12:operation soapAction="urn:addTaskDescriptionInTenant" style="document"/>
+      <wsdl:operation name="taskExists">
+         <soap12:operation soapAction="urn:taskExists" style="document"/>
          <wsdl:input>
             <soap12:body use="literal"/>
          </wsdl:input>
@@ -128,6 +163,15 @@
             <soap12:fault use="literal" name="TenantTaskAdminServiceTaskManagementException"/>
          </wsdl:fault>
       </wsdl:operation>
+      <wsdl:operation name="addTaskDescriptionInTenant">
+         <soap12:operation soapAction="urn:addTaskDescriptionInTenant" style="document"/>
+         <wsdl:input>
+            <soap12:body use="literal"/>
+         </wsdl:input>
+         <wsdl:output>
+            <soap12:body use="literal"/>
+         </wsdl:output>
+      </wsdl:operation>
    </wsdl:binding>
    <wsdl:binding name="TenantTaskAdminServiceHttpBinding" type="ns:TenantTaskAdminServicePortType">
       <http:binding verb="POST"/>
@@ -140,6 +184,15 @@
             <mime:content type="text/xml" part="parameters"/>
          </wsdl:output>
       </wsdl:operation>
+      <wsdl:operation name="taskExists">
+         <http:operation location="taskExists"/>
+         <wsdl:input>
+            <mime:content type="text/xml" part="parameters"/>
+         </wsdl:input>
+         <wsdl:output>
+            <mime:content type="text/xml" part="parameters"/>
+         </wsdl:output>
+      </wsdl:operation>
       <wsdl:operation name="addTaskDescriptionInTenant">
          <http:operation location="addTaskDescriptionInTenant"/>
          <wsdl:input>
diff --git a/modules/services/org.wso2.intcloud.services.tenant.carbonapps.deployer/src/main/java/org/wso2/intcloud/services/tenant/carbonapps/deployer/TenantCarbonAppDeployer.java b/modules/services/org.wso2.intcloud.services.tenant.carbonapps.deployer/src/main/java/org/wso2/intcloud/services/tenant/carbonapps/deployer/TenantCarbonAppDeployer.java
index 436d831..370cc43 100644
--- a/modules/services/org.wso2.intcloud.services.tenant.carbonapps.deployer/src/main/java/org/wso2/intcloud/services/tenant/carbonapps/deployer/TenantCarbonAppDeployer.java
+++ b/modules/services/org.wso2.intcloud.services.tenant.carbonapps.deployer/src/main/java/org/wso2/intcloud/services/tenant/carbonapps/deployer/TenantCarbonAppDeployer.java
@@ -38,6 +38,7 @@ public class TenantCarbonAppDeployer extends CarbonAppUploader {
      *
      * @param tenantId              tenant id
      * @param carbonApplicationName Name of the carbon application
+     * @param carbonApplicationPath path to carbon application
      * @return true if the operation successfully completed.
      * @throws AxisFault Thrown if and error occurs while uploading the sample
      */
diff --git a/modules/services/org.wso2.intcloud.services.tenant.tasks/src/main/java/org/wso2/intcloud/services/tenant/tasks/TenantTaskAdminService.java b/modules/services/org.wso2.intcloud.services.tenant.tasks/src/main/java/org/wso2/intcloud/services/tenant/tasks/TenantTaskAdminService.java
index 0ce33f8..7065273 100644
--- a/modules/services/org.wso2.intcloud.services.tenant.tasks/src/main/java/org/wso2/intcloud/services/tenant/tasks/TenantTaskAdminService.java
+++ b/modules/services/org.wso2.intcloud.services.tenant.tasks/src/main/java/org/wso2/intcloud/services/tenant/tasks/TenantTaskAdminService.java
@@ -31,7 +31,7 @@ public class TenantTaskAdminService extends CarbonTaskManagementService {
 
     private static final Log log = LogFactory.getLog(TenantTaskAdminService.class);
 
-    public boolean addTaskDescriptionInTenant(int tenantId, String taskElement) throws TaskManagementException {
+    public boolean addTaskDescriptionInTenant(int tenantId, String taskElement) {
         try {
             PrivilegedCarbonContext.startTenantFlow();
             PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(tenantId, true);
@@ -39,7 +39,11 @@ public class TenantTaskAdminService extends CarbonTaskManagementService {
             return addTaskDescription(AXIOMUtil.stringToOM(taskElement));
 
         } catch (XMLStreamException e) {
-            throw new TaskManagementException(e.getMessage(), e);
+            log.error(e.getMessage(),e);
+            return false;
+        } catch (TaskManagementException e) {
+            log.error(e.getMessage(), e);
+            return false;
         } finally {
             PrivilegedCarbonContext.endTenantFlow();
         }
@@ -59,4 +63,19 @@ public class TenantTaskAdminService extends CarbonTaskManagementService {
         }
     }
 
+    public boolean taskExists(int tenantId, String name, String group)
+            throws TaskManagementException {
+        log.info("Getting task " + name);
+        try {
+            PrivilegedCarbonContext.startTenantFlow();
+            PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(tenantId, true);
+
+            OMElement taskDescription = getTaskDescription(name, group);
+
+            return (taskDescription != null);
+
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
 }
\ No newline at end of file
diff --git a/pom.xml b/pom.xml
index f81f7f5..60a6907 100644
--- a/pom.xml
+++ b/pom.xml
@@ -121,6 +121,11 @@
             </dependency>
             <dependency>
                 <groupId>org.wso2.appcloud</groupId>
+                <artifactId>org.wso2.intcloud.services.tenant.ntasks.stub</artifactId>
+                <version>${appcloud.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.appcloud</groupId>
                 <artifactId>org.wso2.intcloud.common</artifactId>
                 <version>${appcloud.version}</version>
             </dependency>
@@ -227,6 +232,11 @@
                 <artifactId>org.wso2.carbon.task</artifactId>
                 <version>${carbon.mediation.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.mediation</groupId>
+                <artifactId>org.wso2.carbon.mediation.ntaskadmin</artifactId>
+                <version>${carbon.mediation.version}</version>
+            </dependency>
         </dependencies>
     </dependencyManagement>
 
@@ -275,11 +285,13 @@
         <module>modules/services/org.wso2.intcloud.services.tenant.carbonapps.undeployer</module>
         <module>modules/services/org.wso2.intcloud.services.tenant.templates</module>
         <module>modules/services/org.wso2.intcloud.services.tenant.tasks</module>
+        <module>modules/services/org.wso2.intcloud.services.tenant.ntasks</module>
         <module>modules/service-stubs/org.wso2.intcloud.services.tenant.carbonapps.deployer.stub</module>
         <module>modules/service-stubs/org.wso2.intcloud.services.tenant.carbonapps.undeployer.stub</module>
         <module>modules/service-stubs/org.wso2.intcloud.services.tenant.templates.stub</module>
         <module>modules/service-stubs/org.wso2.intcloud.services.tenant.tasks.stub</module>
-        
+        <module>modules/service-stubs/org.wso2.intcloud.services.tenant.ntasks.stub</module>
+
         <!--module>modules/integreation/integration-tests</module-->
     </modules>
 
@@ -360,7 +372,7 @@
     </issueManagement>
 
     <properties>
-        <carbon.platform.version>4.2.0</carbon.platform.version>
+        <carbon.platform.version>4.4.5</carbon.platform.version>
         <equinox.osgi.version>3.10.2.v20150203-1939</equinox.osgi.version>
         <appcloud.version>3.0.0-SNAPSHOT</appcloud.version>
         <equinox.osgi.services.version>3.4.0.v20140312-2051</equinox.osgi.services.version>
